Optimize Performance with Caching for Resolved Options and Parsed Locale Strings

nodify-at opened the pull request:
#### Changes
* **Caching:** Implemented caching to avoid redundant computations of resolved options and locale parsing.
* **Efficient Checks:** Introduced a final isEnglish variable to efficiently determine if the locale supports English.
* **Initialization**: supportsFastNumbers is now immediately initialized to prevent unnecessary calls.

#### Performance Impact
Tests indicate up to a 10x speed improvement with these enhancements.

#### Issue Fixed
Fixes #1428

nodify-at commented:
Is there any update on this or can someone give me a feedback if I've to improve this PR? :)

flodlc commented:
Once this one is merged, could we expect a new release ? There has been many perf improvements that would benefits to the community.

icambron commented:
I'll take a look at this one soon

nodify-at commented:
Thanks @icambron :)

Seledrex commented:
Nice work! I was just investigating this problem myself and stumbled upon this issue. Definitely looking forward to this performance improvement! 🚀

icambron commented on src/impl/locale.js:
```diff
@@ -374,14 +394,14 @@ export default class Locale {
 
     this.specifiedLocale = specifiedLocale;
-    this.fastNumbersCached = null;
+    this.fastNumbersCached = supportsFastNumbers(this);
```
I guess I don't quite know what moving this here is fixing

icambron commented on src/impl/locale.js:
```diff
@@ -374,14 +394,14 @@ export default class Locale {
     this.specifiedLocale = specifiedLocale;
-    this.fastNumbersCached = null;
+    this.fastNumbersCached = supportsFastNumbers(this);
+    this.isEnglishCached =
```
Similarly, I agree we should cache this, but why not cache it inside `isEnglish`. That way if `isEnglish` is never called, we don't pay the performance penalty

icambron commented on src/impl/locale.js:
```diff
@@ -123,10 +148,9 @@ function intlConfigString(localeStr, numberingSystem, outputCalendar) {
-  } else {
-    return localeStr;
   }
+  intlConfigCache[key] = localeStr;
```
I'm having a hard time understanding how this would give such a big speedup. Can you use the benchmarks in the repo to demonstrate this?

icambron commented on src/impl/locale.js:
```diff
@@ -73,7 +81,18 @@ function getCachedWeekInfo(locString) {
 
-function parseLocaleString(localeStr) {
+const localeStringCache = {};
+function getParsedCachedLocaleString(localeStr) {
```
I can believe that this parsing is slow, but I'd like to see the benchmarks on this too

icambron reviewed (changes requested):
Sorry for taking so long to review this. I left some comments

nodify-at commented:
@icambron Thanks for your feedback. I took a thorough approach initially to improve the calls, but I've now simplified it to focus on the isEnglish and fastNumbers calls only. I've also cleaned up all other improvements/changes. Could you please review it again?

nodify-at commented:
@icambron 
I hope I did this correctly. I also added two new benchmark tests to compare the results.

Here are the initial results from my local environment:

DateTime#format in German: 301,824 ops/sec ±0.41% (95 runs sampled)
DateTime#format in German with no-cache: 11,999 ops/sec ±3.03% (88 runs sampled)

nodify-at commented:
@icambron did you get my last changes ?

flodlc commented:
I'm starting to lose hope that this PR will ever get merged 😞

Seledrex commented:
> I'm starting to lose hope that this PR will ever get merged 😞 

Just give it some more time!

icambron commented:
Looks good, merged. Thanks for the contribution. I will release this soon.

nodify-at commented:
Awesome :) Thanks @icambron

mohd-akram commented on src/impl/locale.js:
```diff
@@ -167,7 +175,7 @@ function supportsFastNumbers(loc) {
       !loc.locale ||
       loc.locale.startsWith("en") ||
-      new Intl.DateTimeFormat(loc.intl).resolvedOptions().numberingSystem === "latn"
+      getCachedIntResolvedOptions(loc.locale).numberingSystem === "latn"
```
Should this be `loc.intl`?

nodify-at replied:
It looks a bit strange, because in my last commit I reverted this change. getCachedIntResolvedOptions function is used only in supportsFastNumbers and isEnglish .

See the code in master: https://github.com/moment/luxon/blob/master/src/impl/locale.js

nodify-at replied:
Ops, I see your pointed code better now, sorry and you're right it should be intl.

nodify-at replied:
https://github.com/moment/luxon/pull/1654
